clippy_lints/src/missing_fields_in_debug.rs

```diff
@@ -0,0 +1,293 @@
+const HELP_INCLUDE_ALL_FIELDS: &str = "consider including all fields in this `Debug` impl";
+const HELP_FINISH_NON_EXHAUSTIVE: &str = "consider calling `.finish_non_exhaustive()` if you intend to ignore fields";
+
+fn report_lints(cx: &LateContext<'_>, span: Span, span_notes: Vec<(Span, &'static str)>) {
+    span_lint_and_then(cx, MISSING_FIELDS_IN_DEBUG, span, WARN, |diag| {
+        for (span, note) in span_notes {
+            diag.span_note(span, note);
+        }
+        diag.help(HELP_INCLUDE_ALL_FIELDS).help(HELP_FINISH_NON_EXHAUSTIVE);
+    });
+}
+
+/// Checks if we should lint in a block of code
+///
+/// The way we check for this condition is by checking if there is
+/// a call to `Formatter::debug_struct` or `Formatter::debug_tuple`
+/// and no call to `.finish_non_exhaustive()`.
+fn should_lint<'tcx>(
+    cx: &LateContext<'tcx>,
+    typeck_results: &TypeckResults<'tcx>,
+    block: impl Visitable<'tcx>,
+) -> bool {
+    let mut has_finish_non_exhaustive = false;
+    let mut has_debug_struct_tuple = false;
+
+    let _: Option<!> = for_each_expr(block, |expr| {
+        let ExprKind::MethodCall(path, recv, ..) = &expr.kind else {
+            return ControlFlow::Continue(());
+        };
+        let recv_ty = typeck_results.expr_ty(recv).peel_refs();
+
+        if_chain! {
+            if [sym::debug_struct, sym::debug_tuple].contains(&path.ident.name);
+            if match_type(cx, recv_ty, &paths::FORMATTER);
+            then {
+                has_debug_struct_tuple = true;
+            }
+        }
+
+        if_chain! {
+            if path.ident.name.as_str() == "finish_non_exhaustive";
+            if match_type(cx, recv_ty, &paths::DEBUG_STRUCT);
+            then {
+                has_finish_non_exhaustive = true;
+            }
+        }
+        ControlFlow::Continue(())
+    });
+
+    !has_finish_non_exhaustive && has_debug_struct_tuple
+}
+
+/// Attempts to find unused fields assuming that the item is a struct
+fn check_struct<'tcx>(
+    cx: &LateContext<'tcx>,
+    typeck_results: &TypeckResults<'tcx>,
+    block: &'tcx Block<'tcx>,
+    self_ty: Ty<'tcx>,
+    item: &'tcx Item<'tcx>,
+    data: &VariantData<'_>,
+) {
+    let mut field_accesses = FxHashSet::default();
+
+    let _: Option<!> = for_each_expr(block, |expr| {
+        if_chain! {
+            if let ExprKind::Field(target, ident) = expr.kind;
+            let target_ty = typeck_results.expr_ty(target).peel_refs();
+            if target_ty == self_ty;
+            then {
+                field_accesses.insert(ident);
+            }
+        }
+        ControlFlow::Continue(())
+    });
+
+    let span_notes = data
+        .fields()
+        .iter()
+        .filter_map(|field| {
+            if field_accesses.contains(&field.ident) {
+                None
+            } else {
+                Some((field.span, HELP_FIELD_UNUSED))
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if !span_notes.is_empty() {
```

Comment:
We probably want to check that at least one field is being accessed, so the patterns like the following aren't linted

```rust
use std::fmt::{Debug, Formatter, Result};

struct Inner {
    a: usize,
    b: usize,
}

struct HasInner {
    inner: Inner,
}

impl HasInner {
    fn get(&self) -> &Inner {
        &self.inner
    }
}

impl Debug for HasInner {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        let inner = self.get();

        f.debug_struct("HasInner")
            .field("a", &inner.a)
            .field("b", &inner.b)
            .finish()
    }
}
```

Plus a test for that